src/shogun-core-main/src/main/java/de/terrestris/shoguncore/util/interceptor/secure/WmsRequestInterceptor.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetLegendGraphic(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WMS GetLegendGraphic request");
+        return isAllowed(request, "layer") ? new GeoserverAuthHeaderRequest(
+            request, gsUser, gsPass) : forbidRequest(request);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetStyles(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WMS GetStyles request");
+        return isAllowed(request) ? new GeoserverAuthHeaderRequest(
+            request, gsUser, gsPass) : forbidRequest(request);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetCapabilities(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WMS GetCapabilities request");
+        // response will be intercepted
+        return new GeoserverAuthHeaderRequest(request, gsUser, gsPass);
+    }
+
+    /**
+     * Check for read permission on layer by getting it through service
+     *
+     * @param request The request
+     * @param paramName The optional parameter name for the layer parameter
+     * @return if the layer is allowed to read for current user
+     */
+    private boolean isAllowed(MutableHttpServletRequest request,
+        String paramName) {
+        String layersParam = request.getParameterIgnoreCase(paramName);
+        List<Layer> all = layerService.findAll();
+        boolean match = false;
+        for (Layer layer : all) {
+            if (layer.getSource() instanceof ImageWmsLayerDataSource) {
+                ImageWmsLayerDataSource source = (ImageWmsLayerDataSource)
+                    layer.getSource();
+                if (source.getLayerNames().equalsIgnoreCase(layersParam) &&
+                    source.getUrl().equalsIgnoreCase(request.getContextPath() +
+                    "/geoserver.action")) {
+                    match = true;
+                }
+            }
+        }
+        return match;
+    }
+
+    /**
+     * Calls main method with default "layers" parameter
+     *
+     * @param request The request
+     * @return If the layer is allowed to read for current user
+     */
+    private boolean isAllowed(MutableHttpServletRequest request) {
+        return isAllowed(request, "layers");
+    }
+
+    /**
+     * <p>forbidRequest.</p>
+     *
+     * @param request a {@link de.terrestris.shoguncore.util.interceptor.MutableHttpServletRequest} object.
+     * @return a {@link de.terrestris.shoguncore.util.interceptor.MutableHttpServletRequest} object.
+     */
+    protected MutableHttpServletRequest forbidRequest(MutableHttpServletRequest request) {
+        this.setAppUriFromRequest(request);
+        String redirectUri = appUri == null ? "" : appUri.toString();
+        request.setRequestURI(redirectUri + "/response/forbidden.action");
+        return request;
+    }
```

Comment:
As this method is present in several interceptors a common abstract class containg this one could be helpful.